### DISCUSSION TITLE
Can I create a unique ID for Public User?

### DISCUSSION THREAD

**Discussion by BKDragon** [Budibase/budibase#11006] [Help] [answered]
I'm trying to add analysis capabilities to my Budibase App(base on Budibase Cloud), where I can record user behavior by triggering a query for each action.
But the question is, how do I know which behaviors are triggered by the same user. It's easy for registered users, but I want to track those Public User.

**Comment by ConorWebb96** [comment]
Hey @BKDragon,

You likely could do something with the `Update state` action. Where you Randomly generate a unique ID and set it on the loading of the page. From here you could use this state to uniquely identify public users as it could be passed in as the id for forms etc. This all assumes that these users use your app with the same device every-time as I believe is stored at the device level. Not 100% on how long persisted state can be stored either but its worth looking at this!

I hope this helps!

**Reply by BKDragon** [reply]
I can be born into a public user-generated random ID, but I'm not sure how to save the only ID I generated.
Writing to cookie or localStorage does not seem to be supported in the embed component.
If I can't save this unique ID and make sure that the ID is always loaded globally when the same user visit, I can't keep track of a single user permanently.

**Reply by ConorWebb96** [reply, marked as answer]
Hey @BKDragon,

Due to this not being something you can natively do in Budibase. I've had to hack some things together to get this functionality working. I would recommend looking at making a custom plugin to include functionality like this. However, if you don't want to do that you can see the example below of what I've done and from there try to change it to suit your needs.

I won't explain what the code does, as from your mentioning of cookies and localStorage I'm assuming you will already know.

1. Add an embed component to your page.
2. Add the below code to the embed component.
```
<img style="display: none;" src onerror="if (document.cookie.indexOf('uniqueID') === -1) {document.cookie = 'uniqueID={{ random 0 9999999999 }}'; } else { }">
```
3. Create a button
4. Add an `Update state` action to the button and add the following function to it.
```
var cookieName = "uniqueID=";
var decodedCookie = decodeURIComponent(document.cookie);
var cookieArray = decodedCookie.split(';');

for (var i = 0; i < cookieArray.length; i++) {
  var cookie = cookieArray[i].trim();
  if (cookie.indexOf(cookieName) === 0) {
    return cookie.substring(cookieName.length, cookie.length);
  }
}

return 'Nothing'
```
5. Click the preview button.
6. On load the cookie should be saved to then to save the id to state click the button.
<img width="935" alt="Screenshot 2023-06-30 at 12 04 08" src="https://github.com/Budibase/budibase/assets/126772285/77e563b7-b490-4fde-8f9c-3022c2d92c10">
<img width="666" alt="Screenshot 2023-06-30 at 12 04 19" src="https://github.com/Budibase/budibase/assets/126772285/c50ae55b-d395-43e5-96a0-b04d88ac48f9">
<img width="776" alt="Screenshot 2023-06-30 at 12 08 56" src="https://github.com/Budibase/budibase/assets/126772285/9bfddc98-5cb4-44bd-a3b2-fa740da1d330">



This is a simple demo to point you in the right direction. Hopefully this helps.
